staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go

```diff
@@ -143,9 +164,13 @@ func WithPriorityAndFairness(
 				epmetrics.DroppedRequests.WithContext(ctx).WithLabelValues(epmetrics.ReadOnlyKind).Inc()
 			}
 			epmetrics.RecordRequestTermination(r, requestInfo, epmetrics.APIServerComponent, http.StatusTooManyRequests)
+			if isWatchRequest {
+				wg.Done()
+			}
 			tooManyRequests(r, w)
 		}
-
+		// In case of watch, from P&F POV it already finished, but we need to wait until the request itself finishes.
+		wg.Wait()
```

Comment:
Good catch, although this comment is appearing in the wrong place.  The `wg.Wait()` here has nothing to do with the request context's Done channel.  However, in the StartRequest method in queueset.go, there _is_ code with a comment noting that it relies on that channel eventually being closed.  Now, your comment wasn't entirely clear on this point.  In the case of a WATCH, does the request context have a Done channel that is never closed, or does it have no Done channel?

In a similar vein to the dependence on eventual closure of the request context's Done channel, this PR is also introducing dependence on the watch initialization signal eventually coming.  What I meant by my earlier comment about not vouching for the storage part includes this.  I do not know that this dependence is satisfied.  And perhaps it would be better to not depend on that.

Reply:
> We might need to do some plumbing so we have something similar to:

Let me take a look into it. What I'm wondering about is what this doesn't cause any test failures.
Isn't the context actually eventually closed when the request is processed (the same way as for regular requests)?


> This PR is also introducing dependence on the watch initialization signal eventually coming.

This isn't true. See lines 138-140 - they are protecting us from the signal not coming.
[But that defence in depth - this signal will always come if the request will reach the storage layer.]

Reply:
> What I'm wondering about is what this doesn't cause any test failures.
Isn't the context actually eventually closed when the request is processed (the same way as for regular requests)?

what may be happening here is that `ctx.Done() == nil` for `Watch` since we are not setting any deadline for long running and watch requests. (`WithRequestDeadline` only sets deadline to non long running requests).
the goroutine is spun up only when `ctx.Done() ` is non nil, see the code snippets from `StartRequest`
```
        doneCh := ctx.Done()
	if doneCh != nil {
		qs.preCreateOrUnblockGoroutine()
		go func() {
			defer runtime.HandleCrash()
			qs.goroutineDoneOrBlocked()
			_ = <-doneCh
			// Whatever goroutine unblocked the preceding receive MUST
			// have already either (a) incremented qs.counter or (b)
			// known that said counter is not actually counting or (c)
			// known that the count does not need to be accurate.
			// BTW, the count only needs to be accurate in a test that
			// uses FakeEventClock::Run().
			klog.V(6).Infof("QS(%s): Context of request %q %#+v %#+v is Done", configName, fsName, descr1, descr2)
			qs.cancelWait(req)
			qs.goroutineDoneOrBlocked()
		}()
	}
```
so we may not creating the goroutine at all for `Watch`.

Reply:
> isn't the context actually eventually closed when the request is processed

so I just did a quick scan of the filters and looks like for a `Watch` we do not apply any `WithCancel` or `WithTimeout` to the request context. 
The only filter that sets a deadline to a request before the APF filter is `WithRequestDeadline` and it only applies it to non long running requests.

I also checked the net/http2 library, looks like just before the http handler is called, net/http creates a context with `WithCancel` 
```
	ctx, cancelCtx := context.WithCancel(sc.baseCtx)
	st := &http2stream{
		sc:        sc,
		id:        id,
		state:     state,
		ctx:       ctx,
		cancelCtx: cancelCtx,
	}
``
and then creates a new `http.Request` with this context and then invokes the handler. So I think I may have been wrong when I said that `tx.Done() == nil` and we never create that go routine. (let me do some tests to verify that)